### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by zepplondon**
:wave:! GitHub Classroom created this pull request as a place for your teacher to leave feedback on your work. It will update automatically. **Don’t close or merge this pull request**, unless you’re instructed to do so by your teacher.

In this pull request, your teacher can leave comments and feedback on your code. Click the **Subscribe** button to be notified if that happens.

Click the **Files changed** or **Commits** tab to see all of the changes pushed to `master` since the assignment started. Your teacher can see this too.

<details>
<summary>
<strong>Notes for teachers</strong>
</summary>

Use this PR to leave feedback. Here are some tips:

- Click the **Files changed** tab to see all of the changes pushed to `master` since the assignment started. To leave comments on specific lines of code, put your cursor over a line of code and click the blue **+** (plus sign). To learn more about comments, read “[Commenting on a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request)”.
- Click the **Commits** tab to see the commits pushed to `master`. Click a commit to see specific changes.
- If you turned on autograding, then click the **Checks** tab to see the results.
- This page is an overview. It shows commits, line comments, and general comments. You can leave a general comment below.

For more information about this pull request, read “[Leaving assignment feedback in GitHub](/help/leaving-feedback-in-github)”.
</details>

Subscribed: @SonaSaha

### CODE REVIEW

File: src/AbstractProduct.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import java.util.ArrayList;
+
+public abstract class AbstractProduct implements IProduct {
+  protected ArrayList<MaterialStack> payloadProduct = new ArrayList<>();
+  protected ArrayList<MaterialStack> costProduct = new ArrayList<>();
+  protected String category;
```
Comment (azakhary) [inline src/AbstractProduct.java:6]: no need to remember the category inside the product. All you have to do is to create a hashmap in the Shop instead so it knows which product is from what category.
Think of Category of being a Shop "room" or "isle", where they sell Milk stuff for example. The "Cheese" product does not have written on it "I am located in milk section". instead, it is just located in milk section - and milk section belongs to shop. If it makes sense in real world, then it will also make sense in OOP

File: src/AbstractShop.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+public abstract class AbstractShop implements IShop {
+    /**
+     * hashMap contains products and it's unique id
+     */
+     protected HashMap<Integer,IProduct> productMap = new HashMap<>();
+    /**
+     * static member to remember unique id
+     */
+    protected static Integer count = 1;
```
Comment (azakhary) [inline src/AbstractShop.java:12]: why is this field?

File: src/AbstractShop.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        productMap.remove(id);
+    }
+
+    /**
+     * Filters by category
+     * @param category
+     * @return  filtered array
+     */
+    @Override
+    public ArrayList<IProduct> getProductsByCategories (String category) {
```
Comment (azakhary) [inline src/AbstractShop.java:67]: instead of doing this have two hashmaps, one for all products to fetch by ID, and another for all product arrays BY category to fetch by category id.

File: src/AbstractUser.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import java.util.ArrayList;
+
+public class AbstractUser implements IUser {
```
Comment (azakhary) [inline src/AbstractUser.java:3]: IUser is a bad idea here to use as something that can have inventory. Think more global.
In you game many different entities can have the functionality to have an inventory. For example Player can have it, but also in game character my have it. They might be different classes, but they will still implement some similar interfaces. maybe even more. While your user still may implement IUser interface for things like "jump", "die" or other that are specific for all types of users. In this case you are interested in interface that is Meant to provide functionality to work with inventory. It's good to separate this. And re-use later.

Create IInventoryProvider interface something that has methods like getMaterials, subsctract, add and so on.
And then your IUser in this case can be even empty for now, and AbstractUser implements IUser, IInventoryProvider

(two interfaces)

If you put everything in one IUser interface yo uwill never be able to implement it by any other class and one off interfaces make no sense (hena miangamic prosto useri mej kgreir eli et sagh metodner@)

File: src/ShopImpl.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import java.util.ArrayList;
+
+public class ShopImpl extends AbstractShop {
+    /**
+     * Ctor for Shop
```
Comment (azakhary) [inline src/ShopImpl.java:5]: կտոր :D

File: src/MaterialType.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    private String name;
+    /**
+     * Material description
+     */
+    private String description;
+    /**
+     * Icon URL to be downloaded from S3 amazon server
+     */
+    private String iconURL;
```
Comment (azakhary) [inline src/MaterialType.java:17]: even if its a minor detail - probably not worth to copy paste a code, write it again, in your own way to just remember better.

**Review by azakhary — CHANGES_REQUESTED** [review]

**Comment by SonaSaha** [comment]
Thanks a lot for comments

File: src/AbstractShop.java (hunk 2 above)
Comment (SonaSaha) [reply src/AbstractShop.java:12]: the count is for remembering current free id

Comment (SonaSaha) [reply src/AbstractShop.java:12]: That's clear,  thanks

File: secondCommit/src/Main.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        IInventoryProvider user = new UserImpl();
+
+        //User have some initial materials
+        user.addProduct(sugar,8);
+        user.addProduct(apricot,10);
+        user.addProduct(cacao,15);
+
+        if (shop.canBuyProduct(user,2)) {
+
+            shop.buyProduct(user, 2);
```
Comment (Tom-Ski) [inline secondCommit/src/Main.java:57]: Inconsistent spacing

File: secondCommit/src/command/BuyProduct.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * executes following
+     * inventoryProvider buys product with the id from the shop
+     * @param shop
+     * @param inventoryProvider
+     * @param id
+     */
+    @Override
+    public void execute(IShop shop, IInventoryProvider inventoryProvider,int id) {
+     shop.buyProduct(inventoryProvider,id);
```
Comment (Tom-Ski) [inline secondCommit/src/command/BuyProduct.java:20]: Inconsistent indent as with rest of the project

File: secondCommit/src/command/CommandManager.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+ * Command manager class is singleton class and manages  commands and shops
+ * and remember the executed commands history
+ */
+public class CommandManager {
+    //instance is private to ban the creation more than one objects
+    private static CommandManager instance;
+    //history array is for remembering  the executed commands
+    ArrayList<ICommand> history = new ArrayList<>();
+    //cursor for indicating current position in the history array
+    private int cursor = 0;
```
Comment (Tom-Ski) [inline secondCommit/src/command/CommandManager.java:18]: historyCursor would be a better name. Cursor can be applied to many things, and does not tell us that its being used for history position. Although you have documented it, we would have to go back to the top of the file to read what this cursor is for.

File: secondCommit/src/command/CommandManager.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * executes command
+     * @param command
+     * @param shop
+     * @param user
+     * @param id
+     */
+    public void executeCommand (ICommand command, IShop shop, IInventoryProvider user,int id) {
+        command.execute(shop,user,id);
+        if(cursor < history.size()) {
```
Comment (Tom-Ski) [inline secondCommit/src/command/CommandManager.java:40]: Inconsistent and unclear spacing here around the function.

if (cursor < history.size()) { would be cleaner

File: secondCommit/src/command/CommandManager.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        cursor++;
+    }
+
+    /**
+     * cancels the last executed command
+     * @param shop
+     * @param user
+     * @param id
+     */
+    public void undo (IShop shop, IInventoryProvider user,int id) {
```
Comment (Tom-Ski) [inline secondCommit/src/command/CommandManager.java:54]: Spacing in method definition is inconsistent. 
, int id)

File: secondCommit/src/command/ICommand.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package command;
+
+import shop.IShop;
+import user.IInventoryProvider;
+
+/**
+ * every command must to have these methods
+ */
+public interface ICommand
+{
```
Comment (Tom-Ski) [inline secondCommit/src/command/ICommand.java:10]: Brackets belong in the same line, not on a new line in java standards.

File: secondCommit/src/poolable/Pool.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    ArrayList<T> freeObjects = new ArrayList<>();
+    /**
+     * array for used objects
+     */
+    ArrayList<T> usedObjects = new ArrayList<>();
+    /**
+     * the number of objects that the class have primarily, but this number can increase depends
+     * on demand
+     */
+    private static int INITIAL_CAPACITY = 100;
```
Comment (Tom-Ski) [inline secondCommit/src/poolable/Pool.java:22]: This can be final as its a constant.

File: src/AbstractShop.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+     * @return product if user can buy otherwise null
+     */
+    @Override
+    public ArrayList<MaterialStack> buyProduct (Integer id,IUser user) {
+        if (!productMap.containsKey(id)) {
+            return null;
+        }
+        boolean checkExist = false;
+        ArrayList<MaterialStack> cost = productMap.get(id).getCost();
+        for (MaterialStack materialStackForUser :user.getUserMaterials()) {
```
Comment (Tom-Ski) [inline src/AbstractShop.java:27]: Double nested for loops are hard to follow, especially if formatting is not consistent.
